{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### CTD$^2$ Beat AML Challenge\n",
    "##### Subchallenge 1\n",
    "_Data files._ \n",
    "* Genetic: rnaseq.csv, dnaseq.csv \n",
    "* Clinical: clinical_numerical.csv, clinical_categorical.csv, clinical_categorical_legend.csv \n",
    "* Response: aucs.csv, response.csv \n",
    "\n",
    "_Goal._ Predict AUC from the given data.\n",
    "\n",
    "_Steps._ \n",
    "* Read RNA-seq, DNA-seq, AUC datasets \n",
    "* Used DNA-seq for feature selection on RNA-seq \n",
    "* Model AUC using RNA-seq \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "# import sys\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import pandas as pd\n",
    "import scipy.linalg as alg\n",
    "# import matplotlib.pyplot as plt\n",
    "from time import time\n",
    "from tqdm import tqdm\n",
    "from skrebate import ReliefF\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.svm import SVR as SupportVectorRegressor\n",
    "from sklearn.linear_model import ElasticNet\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from scipy.stats import pearsonr, spearmanr\n",
    "\n",
    "## Path & Filename...\n",
    "PATH = os.path.join(os.getenv(\"HOMEPATH\"), \"Google Drive\\\\Study\\\\ECE 5332-009 - Topics in EE, Data Science\\\\BeatAML\\\\\")\n",
    "os.chdir(PATH)\n",
    "\n",
    "## Read data...\n",
    "RNA = pd.read_csv(PATH + \"rnaseq.csv\", header = 0)\n",
    "DNA = pd.read_csv(PATH + \"dnaseq.csv\", header = 0)\n",
    "CLI_NUM = pd.read_csv(PATH + \"clinical_numerical.csv\", header = 0)\n",
    "CLI_CAT = pd.read_csv(PATH + \"clinical_categorical.csv\", header = 0)\n",
    "CLI_CAT_LEG = pd.read_csv(PATH + \"clinical_categorical_legend.csv\", header = 0)\n",
    "\n",
    "AUC = pd.read_csv(PATH + \"aucs.csv\",   header = 0)\n",
    "RESP = pd.read_csv(PATH + \"response.csv\", header = 0)\n",
    "\n",
    "# print(\"Dataset sizes = \\n\", pd.DataFrame([RNA.shape, DNA.shape, AUC.shape], index = [\"RNA\", \"DNA\", \"AUC\"], columns = [\"row\", \"col\"]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Feature selection...\n",
    "variant_count = pd.DataFrame([[vv, (DNA.Hugo_Symbol.to_numpy() == vv).sum()] for vv in DNA.Hugo_Symbol.unique()], columns = [\"gene\", \"counts\"])\n",
    "variant_count = variant_count.iloc[variant_count.counts.argsort()[::-1], :].reset_index(drop = True)\n",
    "print(\"#genes with somatic variants = %d\\n#genes with variant count > 1 = %d\" % (variant_count.shape[0], (variant_count.counts > 1).sum()))\n",
    "# print(\"Top 5 genes =\\n\", variant_count[:5])\n",
    "\n",
    "# Check if all genes are in RNA-seq...\n",
    "all([(gg == RNA.Symbol).sum() > 0 for gg in variant_count.gene])\n",
    "genes, gene_idx, _ = np.intersect1d(RNA.Symbol, variant_count.gene, assume_unique = False, return_indices = True)\n",
    "RNA_filt = RNA.iloc[gene_idx, :].copy();    print(\"#genes used = %d\" % len(genes))\n",
    "\n",
    "## Drug information...\n",
    "drug_info, drug_sample_count = [ ], [ ]\n",
    "for dd in AUC.inhibitor.unique():\n",
    "    drug_info.append((dd, AUC.iloc[(AUC.inhibitor == dd).to_numpy(), :]))\n",
    "    drug_sample_count.append([dd, (AUC.inhibitor == dd).sum()])\n",
    "drug_info, drug_sample_count = dict(drug_info), pd.DataFrame(drug_sample_count, columns = [\"inhibitor\", \"lab_id_count\"])\n",
    "print(\"\\n#inhibitors used = %d\" % len(drug_info))\n",
    "# print(\"Top 5 inhibitors = \\n\", drug_sample_count[:5])\n",
    "\n",
    "gene_var = RNA_filt.iloc[:, 2:].var(axis = 1)\n",
    "RNA_filt2 = RNA_filt.iloc[(gene_var > 0.3).to_numpy(), :]\n",
    "genes2 = genes[(gene_var > 0.3).to_numpy()]\n",
    "\n",
    "print(RNA_filt2.shape)\n",
    "print(genes2.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function definitions...\n",
    "def RF(X_train, y_train, X_test, seed = None):\n",
    "    mdl = RandomForestRegressor(n_estimators = 100, criterion = 'mse', random_state = seed)\n",
    "    mdl.fit(X_train, y_train);     y_pred = mdl.predict(X_test)\n",
    "    return y_pred\n",
    "\n",
    "def SVM(X_train, y_train, X_test, seed):\n",
    "    mdl = SupportVectorRegressor(kernel = \"poly\", degree = 3, gamma = \"auto\", tol = 1e-3)\n",
    "    mdl.fit(X_train, y_train);     y_pred = mdl.predict(X_test)\n",
    "    return y_pred\n",
    "\n",
    "def PerfEval(y, y_hat, corr = \"PCC\", err = \"NRMSE\", alpha = 0.05, return_pval = False):\n",
    "    ## Base functions...\n",
    "    CORR  = lambda y, y_hat, c_type: pearsonr(y, y_hat) if c_type == \"PCC\" else spearmanr(y, y_hat)\n",
    "    ERROR = lambda y, y_hat, e_ord: alg.norm(y - y_hat, ord = e_ord) ** e_ord / y.size\n",
    "    \n",
    "    ## Format input...\n",
    "    y, y_hat = np.array(y, dtype = float), np.array(y_hat, dtype = float)\n",
    "    if y.ndim == 2 and any([ss == 1 for ss in y.shape]):\n",
    "        y = y.squeeze()\n",
    "    elif y.ndim > 2:\n",
    "        print(\"Error! y should be a ndarray of shape (n, ) or (n, 1) or (1, n).\")\n",
    "    ####\n",
    "    if y_hat.ndim == 2 and any([ss == 1 for ss in y_hat.shape]):\n",
    "        y_hat = y_hat.squeeze()\n",
    "    elif y_hat.ndim > 2:\n",
    "        print(\"Error! y_hat should be a ndarray of shape (n, ) or (n, 1) or (1, n).\")\n",
    "    ####\n",
    "    \n",
    "    #### Calculate metrics...\n",
    "    if corr.upper() == \"PCC\" or \"SCC\":\n",
    "        rho, pval = CORR(y, y_hat, c_type = corr.upper())\n",
    "        rho = 0 if pval > alpha else rho\n",
    "        print(\"No significant correlation found!\") if pval > alpha else None\n",
    "    else:\n",
    "        print('Not valid! Use \"PCC\" or \"SCC\" for corr option.')\n",
    "    ####\n",
    "    \n",
    "    if \"MAE\" in err.upper():\n",
    "        eps = ERROR(y, y_hat, e_ord = 1)\n",
    "        if err.upper() == \"NMAE\":\n",
    "            eps /= ERROR(y, y.mean(), e_ord = 1)\n",
    "        else:\n",
    "            print('Invalid option! Use a variation of \"MSE\" or \"MAE\" for err option.')\n",
    "    elif \"MSE\" in err.upper():\n",
    "        eps = ERROR(y, y_hat, e_ord = 2)\n",
    "        if err.upper() == \"RMSE\":\n",
    "            eps = sp.sqrt(eps)\n",
    "        elif err.upper() == \"NRMSE\":\n",
    "            eps = sp.sqrt(eps) / y.std(ddof = 0)\n",
    "        else:\n",
    "            print('Invalid option! Use a variation of \"MSE\" or \"MAE\" for err option.')\n",
    "    else:\n",
    "        print('Invalid option! Use a variation of \"MSE\" or \"MAE\" for err option.')\n",
    "    ####\n",
    "    \n",
    "    return (rho, eps, pval) if return_pval else (rho, eps)\n",
    "####\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drugs = AUC.inhibitor.unique().tolist()\n",
    "\n",
    "CV = KFold(n_splits = 5, shuffle = False, random_state = 0)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Pick drug and build dataset for modeling...\n",
    "drug_picked = list(drug_info.keys())[0];    sample_ids = drug_info[drug_picked][\"lab_id\"]\n",
    "print(\"Drug picked for analysis = %s\\n#samples = %d\" % (drug_picked, len(sample_ids)))\n",
    "\n",
    "RNA_drug = RNA.loc[:, sample_ids];    RNA_drug.index = RNA.Symbol\n",
    "AUC_drug = AUC.iloc[(AUC.inhibitor == drug_picked).to_numpy(), 2]\n",
    "p_top = 100\n",
    "FS = ReliefF(n_features_to_select = p_top, n_neighbors = 20)\n",
    "FS.fit(RNA_drug.T.values, AUC_drug.values)\n",
    "\n",
    "X_data = RNA_drug.iloc[FS.top_features_, :].T\n",
    "y_data = AUC_drug\n",
    "\n",
    "# X_data = pd.DataFrame(RNA_filt2.loc[:, sample_ids].to_numpy().T, index = sample_ids, columns = genes2)\n",
    "# y_data = pd.DataFrame(drug_info[drug_picked][\"auc\"].to_numpy(), index = sample_ids, columns = [\"auc\"])\n",
    "print(\"Analysis dataset size = \\n\", pd.DataFrame([X_data.shape, y_data.shape], index = [\"X\", \"y\"], columns = [\"row\", \"col\"]))\n",
    "\n",
    "## Perform prediction...\n",
    "models = [\"RF\", \"SVM\"];    n_sample, n_model = len(sample_ids), len(models)\n",
    "Y_data = pd.DataFrame(dtype = float, index = sample_ids, columns = [\"Actual\"] + models)\n",
    "Y_data[\"Actual\"] = y_data\n",
    "CV = KFold(n_splits = 5, shuffle = False, random_state = None)\n",
    "for train_idx, test_idx in tqdm(CV.split(X_data)):\n",
    "    train_id, test_id = sample_ids.iloc[train_idx], sample_ids.iloc[test_idx]\n",
    "    X_train, y_train = X_data.loc[train_id, :], y_data.iloc[train_idx]\n",
    "    X_test,  y_test  = X_data.loc[test_id, :],  y_data.iloc[test_idx]\n",
    "    \n",
    "    Y_data.loc[test_id, \"RF\"]  =  RF(X_train, y_train, X_test, seed = None)\n",
    "    Y_data.loc[test_id, \"SVM\"] = SVM(X_train, y_train, X_test, seed = None)\n",
    "####\n",
    "\n",
    "metrics = [\"SCC\", \"NRMSE\"]\n",
    "PERF = pd.DataFrame(dtype = float, index = metrics, columns = models)\n",
    "for mm in models:\n",
    "    PERF.loc[:, mm] = PerfEval(y_data, Y_data[mm], corr = metrics[0], err = metrics[1], return_pval = False)\n",
    "####\n",
    "print(\"Model performances for 5-fold cross validation = \\n\", PERF)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_top = 10\n",
    "X_data = RNA_drug.iloc[FS.top_features_[:p_top], :].T\n",
    "y_data = AUC_drug\n",
    "\n",
    "# X_data = pd.DataFrame(RNA_filt2.loc[:, sample_ids].to_numpy().T, index = sample_ids, columns = genes2)\n",
    "# y_data = pd.DataFrame(drug_info[drug_picked][\"auc\"].to_numpy(), index = sample_ids, columns = [\"auc\"])\n",
    "# print(\"Analysis dataset size = \\n\", pd.DataFrame([X_data.shape, y_data.shape], index = [\"X\", \"y\"], columns = [\"row\", \"col\"]))\n",
    "\n",
    "## Perform prediction...\n",
    "models = [\"RF\", \"SVM\"];    n_sample, n_model = len(sample_ids), len(models)\n",
    "Y_data = pd.DataFrame(dtype = float, index = sample_ids, columns = [\"Actual\"] + models)\n",
    "Y_data[\"Actual\"] = y_data\n",
    "CV = KFold(n_splits = 5, shuffle = False, random_state = None)\n",
    "for train_idx, test_idx in tqdm(CV.split(X_data)):\n",
    "    train_id, test_id = sample_ids.iloc[train_idx], sample_ids.iloc[test_idx]\n",
    "    X_train, y_train = X_data.loc[train_id, :], y_data.iloc[train_idx]\n",
    "    X_test,  y_test  = X_data.loc[test_id, :],  y_data.iloc[test_idx]\n",
    "    \n",
    "    Y_data.loc[test_id, \"RF\"]  =  RF(X_train, y_train, X_test, seed = None)\n",
    "    Y_data.loc[test_id, \"SVM\"] = SVM(X_train, y_train, X_test, seed = None)\n",
    "####\n",
    "\n",
    "metrics = [\"SCC\", \"NRMSE\"]\n",
    "PERF = pd.DataFrame(dtype = float, index = metrics, columns = models)\n",
    "for mm in models:\n",
    "    PERF.loc[:, mm] = PerfEval(y_data, Y_data[mm], corr = metrics[0], err = metrics[1], return_pval = False)\n",
    "####\n",
    "print(\"Model performances for 5-fold cross validation = \\n\", PERF)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %qtconsole --style monokai\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
